### PR TITLE
QuickCheck ContractModel testing facilities to deal with dynamically created tokens

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
In this PR we introduce dynamically created tokens to the `ContractModel` testing facilities for `QuickCheck`. We also introduce a `ContractModel` of the `Uniswap` contract and fix some minor issues in and update the existing models for contracts in `plutus-use-cases`.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher Do you mind taking a look at this?

File: plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs [hunk 1]
```diff
@@ -354,10 +414,14 @@ class ( Typeable state
     --   >      Seller :: ContractInstanceKey MyModel MyObsState MySchema MyError
     data ContractInstanceKey state :: * -> Row * -> * -> *
 
+    -- | Get the wallet that the contract running at a specific `ContractInstanceKey` should run
+    -- in
+    instanceWallet :: ContractInstanceKey state a b c -> Wallet
```
Comment (sjoerdvisscher) [inline plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:419]:
```suggestion
    instanceWallet :: ContractInstanceKey state w s e -> Wallet
```

File: plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs [hunk 2]
```diff
@@ -586,24 +689,19 @@ handle handles key =
...
-data ContractMonadState state = ContractMonadState (EmulatorAction state) [ContractInstanceSpec state]
-
-instance Semigroup (ContractMonadState state) where
-    ContractMonadState f xs <> ContractMonadState g ys = ContractMonadState (f <> g) (xs <> ys)
-
-instance Monoid (ContractMonadState state) where
-    mempty = ContractMonadState mempty mempty
 
-newtype EmulatorAction state = EmulatorAction { runEmulatorAction :: Handles state -> EmulatorTrace (Handles state) }
+-- TODO: Refactor this(!)
```
Comment (sjoerdvisscher) [inline plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:702]: Just checking that you didn't leave this here by accident.

**Review by sjoerdvisscher — APPROVED** [review]

File: plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs (hunk 2 above)
Comment (MaximilianAlgehed) [reply plutus-contract/src/Plutus/Contract/Test/ContractModel/Internal.hs:702]: Unfortunately this is here on purpose. Our ambition is to refactor parts of the `ContractModel` machinery but that should come in a future PR when we get around to it - right now it's low priority.

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher I think this is ready to be merged now :)

**Pull request merged**